### PR TITLE
Add lightmap demo

### PR DESCRIPTION
**Pull request by iche033**
depends on:
* ~~osrf/sdformat#429~~ released in 10.1.0
* ~~ignitionrobotics/ign-rendering#182~~
* ~~ignitionrobotics/ign-common#132~~ released in 3.9.0
* ~~ignitionrobotics/ign-fuel-tools#146~~ released in 5.1.1

This PR demo's the new lightmap feature added in ign-rendering. It reads a new lightmap parameter from SDF and creates a material with lightmap using the new ign-rendering APIs. 

One caveat is in the sdf-to-msgs conversion functions. I didn't add a lightmap field to the material.proto file in ign-msgs as that breaks ABI. So it's temporarily stored to the header. We can consider bumping ign-msgs if this is too much of a hack.

An indoor environment model has been uploaded onto [Fuel](https://app.ignitionrobotics.org/OpenRobotics/fuel/models/Indoor%20Lightmap). To run the demo:

```
ign gazebo -v 4 lightmap.sdf
```

indoor_lightmap

Signed-off-by: Ian Chen <ichen@osrfoundation.org>

### CODE REVIEW

**Comment by iche033** [comment]
> I'm not sure exactly where the problem lies, but I get this (followed by maybe a dozen similar errors) when running it:
oops, I've fixed the paths in model.sdf and have updated the model on Fuel

**Comment by chapulina** [comment]
I think you also need to add `light_map` to `ign-fuel-tools` here:

https://github.com/ignitionrobotics/ign-fuel-tools/blob/ef6c20d604fd744d709d1fb7a7e8bbf7a58e5a80/src/LocalCache.cc#L643-L658

Otherwise it's not getting converted to the URL, i.e.

```
                    <pbr>
                        <metal>
                            <albedo_map>https://fuel.ignitionrobotics.org/1.0/OpenRobotics/models/Indoor Lightmap/1/files/materials/textures/Wall_Albedo.png</albedo_map>
                            <roughness_map>https://fuel.ignitionrobotics.org/1.0/OpenRobotics/models/Indoor Lightmap/1/files/materials/textures/Wall_Roughness.png</roughness_map>
                            <environment_map>https://fuel.ignitionrobotics.org/1.0/OpenRobotics/models/Indoor Lightmap/1/files/materials/textures/IndoorCubemap.dds</environment_map>
                            <light_map uv_set="1">model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png</light_map>
                            <metalness>0.0</metalness>
                        </metal>
                    </pbr>
```

And results in these errors:

```
[GUI] [Err] [SystemPaths.cc:357] Unable to find file with URI [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
[GUI] [Err] [SystemPaths.cc:444] Could not resolve file [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
[GUI] [Err] [SceneManager.cc:547] Unable to find file [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
[GUI] [Err] [SystemPaths.cc:357] Unable to find file with URI [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
[GUI] [Err] [SystemPaths.cc:444] Could not resolve file [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
[GUI] [Err] [SceneManager.cc:547] Unable to find file [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
[GUI] [Err] [SystemPaths.cc:357] Unable to find file with URI [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
[GUI] [Err] [SystemPaths.cc:444] Could not resolve file [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
[GUI] [Err] [SceneManager.cc:547] Unable to find file [model://indoor_lightmap/materials/textures/LightmapEqualizeHistogram.png]
```

File: examples/worlds/lightmap.sdf [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            <line own="left" target="left"/>
+            <line own="bottom" target="bottom"/>
+          </anchors>
+        </ignition-gui>
+
+        <play_pause>true</play_pause>
+        <step>true</step>
+        <start_paused>true</start_paused>
+        <service>/world/lightmap/control</service>
+        <stats_topic>/world/lightmap/stats</stats_topic>
```
Comment (chapulina) [inline examples/worlds/lightmap.sdf:72]: `service` and `stats_topic` can be omitted, and will default to the world name

File: examples/worlds/lightmap.sdf [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            <line own="right" target="right"/>
+            <line own="bottom" target="bottom"/>
+          </anchors>
+        </ignition-gui>
+
+        <sim_time>true</sim_time>
+        <real_time>true</real_time>
+        <real_time_factor>true</real_time_factor>
+        <iterations>true</iterations>
+        <topic>/world/lightmap/stats</topic>
```
Comment (chapulina) [inline examples/worlds/lightmap.sdf:97]: Same here, `topic` can be omitted

File: src/Conversions.cc [hunk 3]
```diff
@@ -301,6 +301,14 @@ msgs::Material ignition::gazebo::convert(const sdf::Material &_in)
           asFullPath(workflow->EnvironmentMap(), _in.FilePath()));
       pbrMsg->set_emissive_map(workflow->EmissiveMap().empty() ? "" :
           asFullPath(workflow->EmissiveMap(), _in.FilePath()));
+
+      // todo(anyone) add light_map to material.proto
```
Comment (chapulina) [inline src/Conversions.cc:305]: Can we ticket an issue for this?

File: examples/worlds/lightmap.sdf (hunk 2 above)
Comment (iche033) [reply examples/worlds/lightmap.sdf]: removed. b2bd07074036bf77cc729940b37a7ac03b78b5d2

File: examples/worlds/lightmap.sdf (hunk 1 above)
Comment (iche033) [reply examples/worlds/lightmap.sdf]: removed. b2bd07074036bf77cc729940b37a7ac03b78b5d2

File: src/Conversions.cc (hunk 3 above)
Comment (iche033) [reply src/Conversions.cc:305]: ticketed ignitionrobotics/ign-msgs#109

**Comment by iche033** [comment]
> I think you also need to add light_map to ign-fuel-tools here:

ah good catch, created pull request: ignitionrobotics/ign-fuel-tools#146

File: src/Conversions.cc (hunk 3 above)
Comment (chapulina) [reply src/Conversions.cc:305]: This was addressed on https://github.com/ignitionrobotics/ign-msgs/pull/111. We could wait until `ign-gazebo5` is updated to use `ign-msgs7` and use the new field here (https://github.com/ignition-tooling/release-tools/issues/362)

Comment (iche033) [reply src/Conversions.cc:305]: sure sounds good

Comment (chapulina) [reply src/Conversions.cc:305]: I think this is unblocked now

File: examples/worlds/lightmap.sdf [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+-->
+
+<sdf version="1.7">
+  <world name="lightmap">
+    <physics name="1ms" type="ignored">
+      <max_step_size>0.001</max_step_size>
+      <real_time_factor>1.0</real_time_factor>
+    </physics>
+    <plugin
+      filename="libignition-gazebo-physics-system.so"
```
Comment (nkoenig) [inline examples/worlds/lightmap.sdf:16]: Is it possible to remove some of these plugins now that some plugin loading is automatic?

File: src/Conversions_TEST.cc [hunk 5]
```diff
@@ -218,6 +218,7 @@ TEST(Conversions, Material)
   workflow.SetEmissiveMap("emissive_map.png");
   workflow.SetGlossinessMap("dummy_glossiness_map.png");
   workflow.SetSpecularMap("dummy_specular_map.png");
+  workflow.SetLightMap("light_map.png", 1u);
```
Comment (nkoenig) [inline src/Conversions_TEST.cc:221]: Is the `light_map.png` file suppose to exist?

File: src/Conversions_TEST.cc [hunk 6]
```diff
@@ -247,6 +248,11 @@ TEST(Conversions, Material)
   EXPECT_EQ("ambient_occlusion_map.png", pbrMsg.ambient_occlusion_map());
   EXPECT_EQ("dummy_glossiness_map.png", pbrMsg.glossiness_map());
   EXPECT_EQ("dummy_specular_map.png", pbrMsg.specular_map());
+
+  // todo(anyone) add light_map to material.proto
```
Comment (nkoenig) [inline src/Conversions_TEST.cc:252]: remove this line.

File: src/Conversions.cc [hunk 7]
```diff
@@ -338,6 +346,19 @@ sdf::Material ignition::gazebo::convert(const msgs::Material &_in)
     workflow.SetEnvironmentMap(pbrMsg.environment_map());
     workflow.SetAmbientOcclusionMap(pbrMsg.ambient_occlusion_map());
     workflow.SetEmissiveMap(pbrMsg.emissive_map());
+
+    // todo(anyone) add light_map to material.proto
```
Comment (nkoenig) [inline src/Conversions.cc:350]: remove this line.

File: src/Conversions.cc (hunk 3 above)
Comment (nkoenig) [inline src/Conversions.cc:305]: Remove this line.

File: src/Conversions_TEST.cc (hunk 5 above)
Comment (chapulina) [reply src/Conversions_TEST.cc:225]: No, I think this is just testing that the conversion between SDF and messages works fine, without worrying about resolving the file.

File: src/rendering/SceneManager.cc [hunk 8]
```diff
@@ -538,6 +538,22 @@ rendering::MaterialPtr SceneManager::LoadMaterial(
       else
         ignerr << "Unable to find file [" << emissiveMap << "]\n";
     }
+
+    // light map
+    std::string lightMap = workflow->LightMap();
+    if (!lightMap.empty())
+    {
+      std::string fullPath = common::findFile(lightMap);
```
Comment (chapulina) [inline src/rendering/SceneManager.cc:546]: Should we use `asFullPath` like the other maps are using?

Comment (iche033) [reply src/rendering/SceneManager.cc]: yep done. 1659774

File: src/Conversions.cc (hunk 3 above)
Comment (iche033) [reply src/Conversions.cc]: removed. d28afcc

File: src/Conversions.cc (hunk 7 above)
Comment (iche033) [reply src/Conversions.cc]: removed. d28afcc

File: src/Conversions_TEST.cc (hunk 6 above)
Comment (iche033) [reply src/Conversions_TEST.cc]: removed. d28afcc

**Review by chapulina — APPROVED** [review]

**Pull request merged**